hs/CalendarResponse.hs

```diff
@@ -0,0 +1,285 @@
+        halfCourse session1
+            | session1 == "F" = halfFall
+            | otherwise = eventsByTime code sect start end (dates !! 0)
+                                       systemTime winterHolidays
+
+-- | Creates an event for each start/end time and date.
+eventsByTime :: String -- ^ Course code.
+             -> String -- ^ Course section.
+             -> [[String]] -- ^ Start times.
+             -> [[String]] -- ^ End times.
+             -> [(String, String)] -- ^ Start/End dates.
+             -> String -- ^ Current system time.
+             -> [String] -- ^ Holidays dates.
+             -> [[String]]
+eventsByTime code sect start end dates systemTime holidays =
+    map (eventsByDate code sect systemTime holidays)
+        (zip' start end dates)
+
+-- | Join three lists together in tuples of three elements
+zip' :: [a] -> [b] -> [c] -> [(a,b,c)]
+zip' (x:xs) (y:ys) (z:zs) = (x,y,z):zip' xs ys zs
+zip' _ _ _ = []
+
+-- | Creates an event for each start/end time.
+eventsByDate :: String -- ^ Course code.
+             -> String -- ^ Course section.
+             -> String -- ^ Current system time.
+             -> [String] -- ^ Holidays dates.
+             -> ([String], -- ^ Start time
+             [String], -- ^ End time
+             (String, String)) -- ^ Start/End date.
+             -> [String]
+eventsByDate code sect systemTime holidays (start, end, dates) =
+    zip start end >>= eventsStr code sect systemTime dates holidays
+
+-- | Generates the string that represents each event.
+eventsStr :: String -- ^ Course code.
+          -> String -- ^ Course section.
+          -> String -- ^ Current system time.
+          -> (String, String) -- ^ Start/End date.
+          -> [String] -- ^ Holidays dates.
+          -> (String, String) -- ^ Start/End time.
+          -> [String]
+eventsStr code sect systemTime (startDate, endDate) holidays (start, end)
+    | (startDate == "" || endDate == "") = []
+    | otherwise =
+        ["BEGIN:VEVENT", "DTSTAMP:" ++ systemTime,
+         "DTSTART;TZID=America/Toronto:" ++ startDate ++ start,
+         "DTEND;TZID=America/Toronto:" ++ startDate ++ end,
+         "RRULE:FREQ=WEEKLY;UNTIL=" ++ endDate ++ "000000Z"] ++
+         map (\date -> "EXDATE;TZID=America/Toronto:" ++ date ++ start)
+             holidays ++
+        ["ORGANIZER:University of Toronto",
+         "SUMMARY:" ++ code ++ " " ++ sect,
+         "CATEGORIES:EDUCATION", "END:VEVENT"]
+
+-- ** Ordering data
+
+-- | Orders the time fields for each course; which were obtained from the
+-- database
+orderData :: [Time] -> [[[Double]]]
+orderData courseFields = filter (not . null)
+                                (assignDay $ map timeField courseFields)
+
+-- | Organizes the data by the day of the week, assigning a position based on
+-- the day in the time field.
+assignDay :: [[Double]] -> [[[Double]]]
+assignDay lst = [monday, tuesday, wednesday, thursday, friday]
```

Comment:
I changed it. However, I am using the information produced by this function to calculate both times and dates. For this reason I need to return [[[Double]]]. Also, the implementation using groupBy and sortBy looks cleaner this way. Thank you for showing these really nice functions to me.